core/impl/operators/crossovers/AdaptiveProbabilityOfGeneCrossover.py

```diff
@@ -0,0 +1,60 @@
+import numpy as np
+from core.operators import Crossover
+from core.Individual import Individual
+from core.Population import Population
+from core.Representation import Representation
+from core.fitness_function.FitnessFunction import FitnessFunction
+
+
+class AdaptiveProbabilityOfGeneCrossover(Crossover):
+    """
+    A class that implements Adaptive Probability of Gene Crossover (Algorytmy genetyczne: kompendium. T. 1 p. 345)
+
+
+    :param how_many_individuals: Number of individuals in the population.
+    :type how_many_individuals: int
+    :param fitness_function: Fitness function used to evaluate the parents.
+    :type fitness_function: FitnessFunction
+    :param probability: Probability of performing crossover.
+    :type probability: float
+    """
+
+    allowed_representation = [Representation.REAL]
+
+    def __init__(
+        self,
+        how_many_individuals: int,
+        fitness_function: FitnessFunction,
+        probability: float = 0,
+    ):
+        super().__init__(how_many_individuals, probability)
+        self.fitness_function = fitness_function
+
+    def _cross(self, population_parent: Population) -> Population:
+        """
+        Steps:
+        1. Randomly select two parents from the population.
+        2. Calculate the fitness of both parents.
+        3. Compute the probability of selecting genes from the first parent based on fitness.
+        4. For each gene, randomly choose it from either parent based on the calculated probability.
+        5. Create the offspring with the selected genes.
+
+        :param population_parent: The population to perform the crossover operation on.
+        :returns: The offspring.
+        """
+        if len(population_parent) < 2:
+            raise ValueError("Population must contain at least two parents")
```

Comment:
Na ten moment nie robimy takich sprawdzań w trakcie wykonania, zakładamy że to będzie spełnione. Do tego len(population_parent) nie zadziała ponieważ population_parent to obiekt Population a nie tablica.

Reply:
usunięte